text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+
+To uniquely identify an `Endpoint`, this crate proposes the `EndpointId` type, which can be implemented as a wrapper around an `Address`.
+
+```rust
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+struct EndpointId(Address);
+```
+
+Note, that this `EndpointId` should be `Clone` and `Copy`, and requires to implement `Eq`, `PartialEq`, and `Hash`, because it is used as a hashmap key in several instances.
+
+## `Command`
+
+`Command`s are messages that are supposed to be issued by higher level business logic like the protocol layer. They are implemented as an `enum`, which in Rust is one way of expressing a polymorphic type:
+
+```rust
+enum Command {
+    // Adds an endpoint to the list.
+    AddEndpoint { url: Url, ... },
+
+    // Removes an endpoint from the list.
+    RemoveEndpoint { epid: EndpointId, ... },
+
+    // Attempts to connect to an endpoint.
+    Connect { epid: EndpointId, ... },
+
+    // Disconnects from an endpoint.
+    Disconnect { epid: EndpointId, ... },
+
+    // Sends a message to a particular endpoint.
+    SendMessage { epid: EndpointId, bytes: Vec<u8>, ... },
+
+    // Sends a message to multiple endpoints.
+    MulticastMessage { epids: Vec<EndpointId>, bytes: Vec<u8>, ...},
+
+    // Sends a message to all endpoints in its list.
+    BroadcastMessage { bytes: Vec<u8>, ... },
+}
+```
+
+This `enum` makes the things the consumer can do with the crate very explicit and descriptive, and also easily extensible.
+
+### `Event`
+
+Similarly to `Command`s, the different kinds of `Event`s in the system are implemented as an `enum` allowing various types of concrete events being sent over event channels:
+
+```rust
+enum Event {
+    // An endpoint was added to the list.
+    EndpointAdded { epid: EndpointId, ... },
+
+    // An endpoint was removed from the list.
+    EndpointRemoved { epid: EndpointId, ... },
+
+    // A new connection is being set up.
+    NewConnection { ep: Endpoint, ... },
+
+    // A broken connection.
+    LostConnection { epid: EndpointId },
+
+    // An endpoint was successfully connected.
+    EndpointConnected { epid: EndpointId, address: Address, ... },
+
+    // An endpoint was disconnected.
+    EndpointDisconnected { epid: EndpointId, ... },
+
+    // A message was sent.
+    MessageSent { epid: EndpointId, num_bytes: usize },
+
+    // A message was received.
+    MessageReceived { epid: EndpointId, bytes: Vec<u8> },
+
+    // A connection attempt should occur.
+    TryConnect { epid: EndpointId, ... },
+}
+
+```
+
+In contrast to commands though, events are messages created by the system, and those that are considered relevant are published for the consumer to execute custom logic.
+
+## Workers
```

Comment:
I feel like it would be important to have also some code of the workers here. The worker designs and their logic might not be implementation detail only. A good worker-design is essential for this crate. Otherwise it is hidden by the RFC and doing suggestions for improvement will require review of the codebase and therefore more difficult to make.
E.g. How are incoming connections handled, how are reading/writing sides spawned, how get messages sent to different peers, how get messages read-in

Reply:
Yea, I was thinking about doing that. I could put more code into the RFC, but I think implementing the workers is pretty much straight forward and I can't think of many different ways of implementing such a worker. But I'll have a look, and see what I can add! :+1:

Reply:
Perfect, thanks!

Reply:
I'm not entirely convinced of putting some code here 🤔 
Sounds like a lot of implementation details.
At the very least maybe only the main loop but I'm not sure if that helps...

Reply:
let's not put code there for now. I can make updates to the RFC later on, if need be. :smile: